onnx/defs/schema.h

```diff
@@ -822,27 +823,58 @@ class OpSchema final {
     return data_propagation_function_ ? true : false;
   }
 
-  bool HasFunction() const {
-    return function_body_.node_size() > 0;
+  std::vector<int> function_opset_versions() const {
+    std::vector<int> opset_versions;
+    std::map<int, std::shared_ptr<FunctionProto>>::const_iterator it = opset_version_to_function_body_.cbegin();
+    for (; it != opset_version_to_function_body_.cend(); ++it) {
+      opset_versions.push_back(it->first);
+    }
+    return opset_versions;
   }
 
-  OpSchema& FunctionBody(const std::vector<NodeProto>& func_nodes);
+  bool HasFunction() const {
+    return !opset_version_to_function_body_.empty();
+  }
 
-  OpSchema& FunctionBody(const std::vector<NodeProto>& func_nodes, const std::vector<OperatorSetIdProto>& opsets);
+  OpSchema& FunctionBody(const std::vector<NodeProto>& func_nodes, int opset_version = kUninitializedSinceVersion);
 
-  OpSchema& FunctionBody(const char* func_body);
+  OpSchema& FunctionBody(
+      const std::vector<NodeProto>& func_nodes,
+      const std::vector<OperatorSetIdProto>& opsets,
+      int opset_version = kUninitializedSinceVersion);
 
-  OpSchema& FunctionAddOpset(const char* domain, int version);
+  OpSchema& FunctionBody(const char* func_body, int opset_version = kUninitializedSinceVersion);
 
   const FunctionProto* GetFunction() const;
 
+  const FunctionProto* GetFunctionWithOpsetVersion(int opset_version) const;
```

Comment:
Documentation would be useful for this function. In particular, I notice that it returns something based on lower_bound. I am not sure about that for 2 reasons. I think the right thing to do is to additionally check that the candidate (returned by lower_bound) has the same opset-version for every op called within the function-body as the requested opset-version. This requires constructing the function-body first. Maybe easier to discuss this in person, over a call.